abcclassroom/clone.py

```diff
@@ -63,18 +90,24 @@ def clone_student_repos(args):
 
     if materials_dir is None:
         print(
-            "No course_materials directory set in config.yml. Will clone "
-            "repositories but will not copy any assignment files."
+            "Oops! I couldn't find a course_materials directory location "
+            "in your config.yml file. I will just clone all of the student"
+            "repositories. I can't copy any assignment files to a "
+            "course_materials directory."
         )
+    # This is a really large try block and makes it hard to figure out where
+    # things fail. perhaps break out into a few sub try / excepts?
     try:
```

Comment:
@kcranston do you happen to konw the best practices with try/accept / ask for forgiveness vs testing everything frm the start?

Here we have a big try block that ends with a fileNotFound. the problem is that there are a few specific things we want to test for including

1. we could get a `keyError` if  ["github_username"] doesn't exist.  (line 109)
2. `FileNotFoundError` on line 106 

right now in general it's hard to troubleshoot as we don't have enough checks. then ofcourse the try you have for `clone_or_update_repo`

to implement tests should i wrap line 106 with a try/except and then line 109 ?
this will just make the code a bit less readable but follows the ask for forgiveness later approach.
vs testing the file in a block above for exists and has the correct keys.

thank you for any feedback on approach here!

Reply:
```
try:
    with open(roster_filename, newline="") as csvfile:
            reader = csv.DictReader(csvfile)
except FileNotFoundError:
   raise FileNotFoundError:
      "oops"
           # then if the above passes
           try:
               reader = csv.DictReader(csvfile)
               for row in reader:
                   student = row["github_username"]
           except KeyError:
               raise KeyError "message"

```

Reply:
that `catch FileNotFoundError` needs to go at the end of the whole `with` statement. Inside that try block, we can then use other nested try / except blocks to catch things that might be wrong with the file, given that it exists.

Reply:
thank you. maybe we can chat more later. this is not behaving the way my brain thinks it should so i need to adjust my thought process! it's also not throwing a keyerror even when it should. it just fails quietly even if the csv file is totally empty.

Reply:
ok i *think* i got it right now! thank you for the help today @kcranston !!! i tested for a keyerror and it worked.